src/kudu/consensus/routing.cc

```diff
@@ -0,0 +1,488 @@
+    std::unordered_map<std::string, std::unique_ptr<Node>>* forest) {
+
+  unordered_map<string, string> dest_to_proxy_from; // keyed by directed edge destination
+  for (const auto& edge : proxy_topology.proxy_edges()) {
+    if (ContainsKey(dest_to_proxy_from, edge.peer_uuid())) {
+      return Status::InvalidArgument("invalid proxy topology: duplicate destination uuid",
+                                     edge.peer_uuid());
+    }
+    InsertOrDie(&dest_to_proxy_from, edge.peer_uuid(), edge.proxy_from_uuid());
+  }
+
+  // Initially, construct a forest comprised of all peers in the Raft config,
+  // with no proxy_from relationships represented.
+  std::unordered_map<std::string, std::unique_ptr<Node>> tmp_forest;
+  std::unordered_map<std::string, Node*> tmp_index;
+  for (const RaftPeerPB& peer : raft_config.peers()) {
+    unique_ptr<Node> hop(new Node(peer));
+    tmp_index.emplace(peer.permanent_uuid(), hop.get());
+    auto result = tmp_forest.emplace(peer.permanent_uuid(), std::move(hop));
+    if (!result.second) {
+      return Status::InvalidArgument("invalid raft config: duplicate peer uuid",
+                                     peer.permanent_uuid());
+    }
+  }
+
+  // proxy_from nodes specified in ProxyTopologyPB that were not found in RaftConfigPB.
+  vector<string> proxy_from_nodes_not_found;
+
+  // Now, organize the forest into parent-child relationships, where the parent
+  // is represented as proxy_from in each ProxyTopologyPB edge, and the child
+  // is the destination. Any node without a valid proxy_from (either not
+  // specified in ProxyTopologyPB or specified as a peer that isn't currently a
+  // member of the Raft config) will be left as a tree root in the forest.
+  for (const RaftPeerPB& peer : raft_config.peers()) {
+    const string* proxy_from_uuid = FindOrNull(dest_to_proxy_from, peer.permanent_uuid());
+    if (!proxy_from_uuid || proxy_from_uuid->empty()) {
+      continue;
+    }
+
+    // Node has proxy_from set, so we must link them and assign object
+    // ownership as a child of the proxy_from Node.
+    Node* proxy_from_ptr = FindWithDefault(tmp_index, *proxy_from_uuid, nullptr);
+    if (!proxy_from_ptr) {
+      proxy_from_nodes_not_found.push_back(*proxy_from_uuid);
+      continue;
+    }
+
+    // Move destination out of forest map and into the proxy_from Node as a child.
+    const string& node_uuid = peer.permanent_uuid();
+    auto iter = tmp_forest.find(node_uuid);
+    DCHECK(iter != tmp_forest.end());
+    unique_ptr<Node> node = std::move(iter->second);
+    tmp_forest.erase(iter->first);
+    node->proxy_from = proxy_from_ptr;
+    auto result = proxy_from_ptr->children.emplace(node_uuid, std::move(node));
+    DCHECK(result.second) << "invalid config: duplicate uuid: " << node_uuid;
+  }
+
+  *index = std::move(tmp_index);
+  *forest = std::move(tmp_forest);
+
+  // This is just a warning, not an error.
+  if (!proxy_from_nodes_not_found.empty()) {
+    return Status::Incomplete(
```

Comment:
I am assuming that Incomplete is required because routing trees can have extraneous
entities, by dbstatus adding them before adding to ring and removing them later than
actual removal from ring?

Reply:
Yeah exactly, I originally wrote the routing code assuming there can be no extraneous entities, but now it makes much more sense to be as forgiving as possible so that we can make proxying topology changes before making the associated config changes and guarantee routing. Even if we don't do that, it should be supported. I used Incomplete so that we could optionally log a warning while enabling us to log context than we could provide at the lower layers.